### PR TITLE
Update schema for control_origin and implementation_status

### PR DESCRIPTION
**Pull request by jbarnicle**
This PR:
- Retroactively adds support for `implementation_status` to schema versions `2.0.0` and `3.0.0`. 
  - The schema already had this field but masonry did not support it.
- Introduces schema version `3.1.0` as per https://github.com/opencontrol/schemas/issues/25
  - Introduces `control_origins` and `GetControlOrigins()` interface method to Satisfies base interface.
  - Introduces `implementation_statuses` and `GetImplementationStatuses()` interface method to Satisfies base interface

### CODE REVIEW

File: models/components/versions/2_0_0/component_test.go [hunk 1]
```diff
@@ -2,13 +2,14 @@ package component
...
+
+	"github.com/blang/semver"
 	"github.com/opencontrol/compliance-masonry/models/common"
 	"github.com/stretchr/testify/assert"
-	"github.com/blang/semver"
 )
 
 func TestComponentGetters(t *testing.T) {
-	testSatisfies := []Satisfies{{Narrative: "Narrative"}, {}, {}, {}}
+	testSatisfies := []Satisfies{{Narrative: "Narrative"}, {}, {}, {}, {}}
```
Comment (jcscottiii) [inline models/components/versions/2_0_0/component_test.go]: no need for extra `{}`

File: models/components/versions/2_0_0/component_test.go [hunk 2]
```diff
@@ -36,6 +37,7 @@ func TestComponentGetters(t *testing.T) {
 		assert.Equal(t, satisfies.GetParameters(), testSatisfies[idx].GetParameters())
 		assert.Equal(t, satisfies.GetCoveredBy(), testSatisfies[idx].GetCoveredBy())
 		assert.Equal(t, "", satisfies.GetControlOrigin())
+		assert.Equal(t, "", satisfies.GetImplementationStatus())
```
Comment (jcscottiii) [inline models/components/versions/2_0_0/component_test.go]: add 
```go
assert.Equal(t, []string{}, satisfies.GetControlOrigins())
```

File: models/components/versions/3_0_0/component_test.go [hunk 3]
```diff
@@ -41,6 +42,7 @@ func TestComponentGetters(t *testing.T) {
 		}
 		assert.Equal(t, satisfies.GetCoveredBy(), testSatisfies[idx].GetCoveredBy())
 		assert.Equal(t, satisfies.GetControlOrigin(), testSatisfies[idx].GetControlOrigin())
+		assert.Equal(t, satisfies.GetImplementationStatus(), testSatisfies[idx].GetImplementationStatus())
```
Comment (jcscottiii) [inline models/components/versions/3_0_0/component_test.go]: add 
```go
assert.Equal(t, []string{}, satisfies.GetControlOrigins())
```

File: models/components/versions/3_1_0/component_test.go [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+		for i, narrative := range satisfies.GetNarratives() {
+			assert.Equal(t, satisfies.GetNarratives()[i].GetKey(), narrative.GetKey())
+			assert.Equal(t, satisfies.GetNarratives()[i].GetText(), narrative.GetText())
+		}
+		assert.Equal(t, satisfies.GetParameters(), testSatisfies[idx].GetParameters())
+		for i, parameter := range satisfies.GetParameters() {
+			assert.Equal(t, satisfies.GetParameters()[i].GetKey(), parameter.GetKey())
+			assert.Equal(t, satisfies.GetParameters()[i].GetText(), parameter.GetText())
+		}
+		assert.Equal(t, satisfies.GetCoveredBy(), testSatisfies[idx].GetCoveredBy())
```
Comment (jcscottiii) [inline models/components/versions/3_1_0/component_test.go]: add an assert for ImplementationStatus

File: models/components/versions/parse.go [hunk 5]
```diff
@@ -37,12 +40,16 @@ func ParseComponent(componentData []byte, fileName string) (base.Component, erro
 		c := new(v3.Component)
 		err = yaml.Unmarshal(componentData, c)
 		component = c
+	case ComponentV3_1_0.EQ(b.SchemaVersion):
```
Comment (jcscottiii) [inline models/components/versions/parse.go]: Need a test for this: https://codecov.io/gh/opencontrol/compliance-masonry/compare/d140f749617e3f5d8427f4df5e37ad32c57ee96d...2ef2863eee20aad0b7f0f7a2b83aff3990790eac?src=pr#6D6F64656C732F636F6D706F6E656E74732F76657273696F6E732F70617273652E676F-43

File: models/components/versions/3_1_0/component.go [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+}
+
+func (s Satisfies) GetControlOrigin() string {
+	return s.ControlOrigin
+}
+
+func (s Satisfies) GetControlOrigins() []string {
+	controlOrigins := set.New()
+	for i := range s.ControlOrigins {
+		controlOrigins.Add(s.ControlOrigins[i])
```
Comment (jcscottiii) [inline models/components/versions/3_1_0/component.go]: Need a test for this: https://codecov.io/gh/opencontrol/compliance-masonry/compare/d140f749617e3f5d8427f4df5e37ad32c57ee96d...2ef2863eee20aad0b7f0f7a2b83aff3990790eac?src=pr#6D6F64656C732F636F6D706F6E656E74732F76657273696F6E732F335F315F302F636F6D706F6E656E742E676F-115

File: models/components/versions/3_1_0/component_test.go (hunk 4 above)
Comment (jbarnicle) [inline models/components/versions/3_1_0/component_test.go]: done

File: models/components/versions/parse.go (hunk 5 above)
Comment (jbarnicle) [inline models/components/versions/parse.go]: done

File: models/components/versions/3_1_0/component.go (hunk 6 above)
Comment (jbarnicle) [inline models/components/versions/3_1_0/component.go]: done

**Comment by jcscottiii** [comment]
@jbarnicle to make PR in schemas

**Comment by afeld** [comment]
Going forward, let's try and be disciplined about merging the changes to the schema before making any schema changes in Masonry. Sound ok?

**Comment by pburkholder** [comment]
I'm late to the game here, but what is the behavior if both control_origin and control_orgins are specified? Warning or Panic or do they  both get rendered into output YAML?

**Comment by jcscottiii** [comment]
@pburkholder if both are specified, the unique features from both are output.
for example, for a given control:
```yaml
  control_origin: inherited
  control_origins:
     - "inherited"
     - "some other origin"
```

The output available to a given renderer would be
```yaml
- "inherited"
- "some other origin"
```

**Comment by afeld** [comment]
From a schema perspective, I would be fine with "behavior undefined", and leave it up to the implementation what to do with that. That's a smart solution for Masonry, though.

**Pull request merged**